### ISSUE TITLE
Question about training for the embedding network of the discriminator

### ISSUE THREAD

**Issue by SSUHan** [shrubb/latent-pose-reenactment#22]
I'm sincerely thanks for sharing your wonderful work and its code publicly. 

As I searched your code, I have a question about the training logic for the embedding network of the discriminator.

In this line :
https://github.com/shrubb/latent-pose-reenactment/blob/master/discriminators/no_landmarks.py#L156
```python
embed = None
if hasattr(self, 'embed'):
    embed = self.embed(label)

fake_in = fake_rgbs
fake_score_G, fake_features = self.pass_inputs(fake_in, embed)
fake_score_D, _ = self.pass_inputs(fake_in.detach(), embed.detach()) # what I want to discuss! 

real_in = target_rgbs
real_score, real_features = self.pass_inputs(real_in, embed)

data_dict['fake_features'] = fake_features
data_dict['real_features'] = real_features
data_dict['real_embedding'] = embed
data_dict['fake_score_G'] = fake_score_G
data_dict['fake_score_D'] = fake_score_D
data_dict['real_score'] = real_score
```

Is there any specific reason to detach the embedded features when we calculate adversarial loss for discriminator on fake images? 

In addition, only by this way, can we guarantee that the embedder's features will be differently mapped among indices?
I think a trivial solution can exist unless we do not apply a regularization term for those features, e.g., uniformity.

Best,
Junsoo Lee

**Comment by shrubb** [comment]
안녕하십니까, 준수씨.

Good questions. Made me reflect for a while.

As for that `.detach()`, I wandered around the code and couldn't understand why it's there. It feels like the system would still work without it. My only hypothesis is that it was a measure in the "Few-Shot Talking Heads" to weaken the discriminator a little bit, so that generator is given more chances to win. Maybe the [projection discriminator paper](https://arxiv.org/abs/1802.05637) will give you more insights, or you can ask the author of [our base paper](https://arxiv.org/abs/1905.08233).

Now to your second question. In traditional GANs, the discriminator is asked "*does this image look realistic?*". We wanted the discriminator to answer a more complex question, "*does this image look (1) realistic and (2) like a valid frame from the `i`th video?*". That's why we employed the so-called projection discriminator. Its architecture is so that its output (real/fake score) is a dot product of two vectors, one predicted by discriminator's convolutional part and one taken from W. If you initialize W with a trivial solution — say, one vector copied to all rows, — the discriminator won't know which video it's dealing with (i.e. won't know `i`), so it will only be able to measure general realism (i.e. answer question *(1)*).

Now note that the discriminator's matrix (`self.embed`, also known as W) is trained only to help discriminator tell real from synthesized. So the adversarial loss will force discriminator to learn different vectors in W.
